### ISSUE TITLE
[Task 28] Decouple Algorithm Configuration from the Training Flow and Add GDPO

### ISSUE THREAD

**Issue by Ternura143** [redai-infra/Relax#96]
## 1. Repository and Development Baseline

| Item | Details |
|---|---|
| Upstream repository | [`redai-infra/Relax`](https://github.com/redai-infra/Relax) |
| Development repository | [`Ternura143/Relax`](https://github.com/Ternura143/Relax) |
| Baseline commit | `039ce876d25540adad847d4223b4de4722d8f425` |
| Development branch | [`task-28-gdpo-registry`](https://github.com/Ternura143/Relax/tree/task-28-gdpo-registry) |

The development branch is based on the commit above. This task will consolidate algorithm names, runtime capabilities, and implementation entry points into a unified registry, then integrate GDPO through that mechanism while preserving the numerical behavior of existing algorithms.

## 2. Runtime Environment

| Item | Configuration |
|---|---|
| Container image | `ghcr.io/redai-infra/relaxrl:dev-20260715-8325919e` |
| Python | `3.12.3` |
| PyTorch | `2.11.0+cu129` |
| CUDA | `12.9.1` |
| GPU | `1 × NVIDIA A800-SXM4-80GB` |
| CPU | `2 × Intel Xeon Platinum 8350C`, 128 logical CPUs |
| Memory | `1.0 TiB` |

Ray, SGLang, Megatron, TransferQueue, and the remaining dependencies use the versions pinned in this image. Registry, dispatch, and formula tests will run on CPU; the end-to-end validation will use the single-GPU environment above.

## 3. Launch Script and Complete Command

A minimal training example will be added at:

```text
examples/algorithms/gdpo/run-qwen3-0.6B-1xgpu-gsm8k.sh
```

The script will be based on the Qwen3-0.6B single-GPU GRPO quickstart provided by the program. It will use GSM8K, Megatron Bridge, SGLang rollout, single-GPU colocate mode, and ClearML. Each rollout contains 4 prompts with 8 responses per prompt, producing a global batch of 32 responses. Ten rollouts therefore correspond to 10 optimizer steps.

```bash
cd relax
PROJECT_NAME=Relax/dev/task-28-gdpo \
MODEL_DIR=./models \
DATA_DIR=./data \
NUM_ROLLOUT=10 \
ROLLOUT_BATCH_SIZE=4 \
N_SAMPLES=8 \
GLOBAL_BATCH_SIZE=32 \
bash examples/algorithms/gdpo/run-qwen3-0.6B-1xgpu-gsm8k.sh
```

The algorithm-specific arguments added or changed by this example are:

```bash
--advantage-estimator gdpo
--gdpo-reward-keys correctness format
--custom-rm-path examples.algorithms.gdpo.reward.reward_func
```

`--gdpo-reward-keys` selects the rewards used by GDPO in the given order. It requires at least two distinct keys and `n_samples_per_prompt >= 2`. The example reward function will return finite floating-point values for both `correctness` and `format` for every response. Their normalized advantages will be summed with equal weights, matching the default behavior of the official GDPO implementation where all reward weights are 1.

The validation commands will be:

```bash
python -m pytest -q \
  tests/utils/training/test_algorithm_registry.py \
  tests/utils/training/test_algorithm_dispatch.py \
  tests/utils/training/test_gdpo.py
pre-commit run --all-files
```

## 4. Problem Analysis, Scope, and Implementation Plan

The baseline maintains algorithm decisions in several independent locations. `relax/core/registry.py::ALGOS` stores Service role mappings, while `relax/utils/arguments.py` separately defines the choices for `--advantage-estimator`. `relax/utils/utils.py::post_process_rewards` uses another list of algorithm names to select reward preprocessing. The fully asynchronous path in `relax/components/advantages.py::Advantages.compute_advantages_and_returns` and the colocate/hybrid path in `relax/backends/megatron/loss.py::compute_advantages_and_returns` each maintain a complete dispatch branch, followed by another name-based dispatch for GSPO, SAPO, and CISPO policy losses. In addition, `Sample.reward` can already store a dictionary, but `Sample.get_reward_value()` selects only one reward key and therefore cannot represent GDPO's independent normalization of multiple rewards.

A lightweight module, `relax/utils/training/algorithm_registry.py`, will become the single source of truth for algorithm metadata and implementation entry points. Each registration will contain the algorithm name, role profile, reward preprocessor, advantage function, policy-loss function, and capability information such as whether a critic, full log probabilities, or multiple rewards are required. Backend implementations will be referenced through lazily loaded function paths, keeping the registry lightweight and preventing argument parsing from importing Ray Serve, Megatron, or concrete Services. `relax/core/registry.py` will derive the existing Controller role mappings from these registrations, while `relax/utils/arguments.py` will obtain algorithm choices, availability, and capability validation from the same source.

Existing reward, advantage, and policy-loss numerical functions will remain in the training utility layer and will be referenced directly by their registrations. The fully asynchronous and colocate/hybrid call paths will only prepare their existing tensors and runtime context before invoking the same registered implementation. Adding an algorithm will therefore require one registration and its implementation rather than synchronized edits to several algorithm-name lists or duplicated top-level dispatch branches.

GDPO multi-reward processing will be placed in the rollout-to-training transformation stage handled by `post_process_rewards()`. At this point, the code still has access to a complete fixed logical training batch before it is written to TransferQueue, together with each sample's `group_index` and raw reward dictionary. The implementation can therefore perform per-reward within-group standardization, combine the results, and apply batch-wise normalization before storing the final scalar advantage in the existing `rewards` field. The training side will continue to use the existing GRPO path that expands scalar rewards into token-level returns, so the TransferQueue schema will not need to carry Python reward dictionaries.

The batch boundary will be validated explicitly. In the standard colocate path, each transformation receives a complete rollout batch. In the standard fully asynchronous/hybrid path, each transformation receives `global_batch_size // num_iters_per_train_update` responses, matching one consumption unit of the Advantages Service. GDPO will verify prompt-group completeness and the fixed logical batch size before the second normalization. In the single-GPU colocate example, this logical training batch is the 32-response global batch. Unit tests will verify that normalization occurs exactly once before data enters TransferQueue and is partitioned across data-parallel workers.

The main files and responsibilities are:

| File | Change |
|---|---|
| `relax/utils/training/algorithm_registry.py` | Define algorithm specifications, registration/query interfaces, and the canonical algorithm catalog |
| `relax/core/registry.py` | Build and validate existing Service role mappings from algorithm registrations |
| `relax/utils/arguments.py` | Generate estimator choices from the registry and add GDPO reward-key validation |
| `relax/utils/utils.py` | Dispatch reward preprocessing through the registry and implement GDPO multi-reward normalization |
| `relax/components/advantages.py` | Invoke registered advantage implementations in the fully asynchronous path |
| `relax/backends/megatron/loss.py` | Invoke registered advantage and policy-loss implementations in the colocate/hybrid path |
| `relax/utils/types.py` | Extend type definitions for multi-reward inputs |

Scope boundaries: this task does not change existing algorithm formulas, default hyperparameters, or loss definitions. It does not refactor the Controller, Service lifecycle, resource scheduling, or TransferQueue protocol, add third-party dependencies, or extend GDPO to reward weighting, runtime-dynamic reward keys, or full paper-metric reproduction. Combinations that split a fixed logical batch—partial rollout, dynamic global batch, and agentic forced flush—are outside this implementation; selecting GDPO with such a fragmented-transfer configuration will produce an explicit configuration error. The end-to-end example is limited to Qwen3-0.6B, GSM8K, and single-GPU colocate mode. Standard fixed-batch fully asynchronous and hybrid paths will be covered through the shared registry implementation and dispatch tests.

## 5. Baseline, Numerical Definition, and Acceptance Criteria

At the baseline commit, `ALGOS` contains four RL role mappings—GRPO, GSPO, SAPO, and CISPO—plus the SFT mapping, while the CLI maintains seven estimator choices. Reward preprocessing has a separate algorithm-name list, advantage computation has two complete dispatch branches, and policy loss has another set of name-based branches. The two REINFORCE++ implementations are present in the advantage paths but absent from the Service `ALGOS`; PPO appears in the CLI choices but is explicitly rejected during argument validation; SFT resolves a separate key through `loss_type`. The registry will represent these availability states as they are and preserve current runtime behavior instead of turning this structural refactor into a change to historical algorithm support. Existing tests cover only the SFT registry and do not yet test algorithm registration/dispatch, before-and-after numerical equivalence, or the multi-reward GDPO formula. These code facts form the structural baseline for this work.

GDPO will follow the statistical definition used by the official [NVlabs/GDPO](https://github.com/NVlabs/GDPO) implementation. Let \(B\) be the response batch to be trained, \(g(i)\) the prompt group containing response \(i\), and \(K\) the configured set of reward keys. Each reward is first standardized independently within its prompt group:

\[
z_{i,k} =
\frac{r_{i,k}-\mu_{g(i),k}}
{\sigma_{g(i),k}+10^{-4}}
\]

The normalized advantages from all rewards are summed with equal weights:

\[
s_i = \sum_{k \in K} z_{i,k}
\]

The combined value is then normalized again over the current response batch:

\[
A_i =
\frac{s_i-\mu_B}
{\sigma_B+10^{-4}}
\]

The computation will use `float32`. Both within-group and batch statistics will use the sample standard deviation, `torch.std(..., correction=1)`. Centering a constant reward produces zero, so that reward contributes zero within the corresponding group; if all rewards are constant, the final advantage is all zeros. A reward dictionary that is missing a configured key or contains a non-numeric or non-finite value will raise an explicit error identifying the sample and key before normalization begins.

The formula and the \(10^{-4}\) epsilon match the valid numerical-input path of NVlabs/GDPO. The upstream TRL implementation first converts `None` to NaN and then replaces NaN with zero through `torch.nan_to_num()`. This implementation will use a stricter input contract so missing and non-finite rewards can be located at Relax's reward boundary instead of silently entering the statistics as zeros.

Registry and dispatch tests will verify that algorithm names, capabilities, and implementation entry points come from one registration source, and that duplicate registrations and unknown names produce clear errors. GDPO will be integrated entirely through its registration. A code-level check will also verify that the new implementation does not depend on `if algorithm == "gdpo"` and that the CLI, role mappings, reward preprocessing, advantage calculation, and policy loss are all resolved through the registry.

Compatibility for existing algorithms will be checked against independent golden tensors generated from the baseline commit. The development branch will use the same CPU tensors, dtype, random seed, and masks to compare GRPO, GSPO, SAPO, CISPO, and both REINFORCE++ advantage implementations. Tests will cover reward normalization on and off, GRPO standard-deviation normalization on and off, zero and non-zero KL, different response lengths, GSPO sequence KL, SAPO, CISPO, and the default policy loss. Existing GRPO reward normalization will continue to use `1e-6`; it will not be changed to GDPO's `1e-4`. All corresponding outputs must satisfy:

```python
torch.allclose(before, after, atol=1e-7, rtol=1e-6)
```

GDPO formula tests will use an independent reference implementation and cover at least two reward keys, multiple prompt groups, different reward scales, collapse of one or all rewards, missing rewards, non-numeric rewards, non-finite values, invalid group cardinality, and zero variance. For valid inputs, outputs must match the reference with `atol=1e-6, rtol=1e-6`, and every output tensor must contain only finite values. Input-validation cases must raise errors that identify the corresponding error type and field. The standard fully asynchronous and colocate/hybrid wrappers must invoke the same registered implementation and produce the same result for the same logical batch. All CPU tests above and `pre-commit run --all-files` must pass.

End-to-end acceptance will use the fixed environment and command from Sections 2 and 3. One run will generate 320 responses, complete 10 optimizer steps, and exit with code 0. The log will include `correctness`, `format`, GDPO advantage, loss, KL, and gradient norm; all numerical metrics must be finite, and generated, enqueued, and trained sample counts must agree. At completion, the Ray Job status must be `SUCCEEDED`, and GPU memory, related processes, and listening ports must return to their pre-run state. Acceptance focuses on functional correctness, numerical equivalence for existing algorithms, and a runnable end-to-end example.

## 6. Delivery

The Pull Request will contain the unified algorithm registration and dispatch module, migration of existing algorithms, the GDPO implementation, CPU unit tests, a minimal `correctness + format` training example, and a concise guide for adding another algorithm. It will also include the complete test commands, test output, and a sanitized end-to-end training log.
